File: packages/utils/src/circuit-breaker/circuit-breaker.ts
// ─── Types ────────────────────────────────────────────────────────────────────

export enum CircuitState {
  CLOSED = 'CLOSED',
  OPEN = 'OPEN',
  HALF_OPEN = 'HALF_OPEN',
}

export interface CircuitBreakerConfig {
  /** Number of consecutive failures before the circuit opens. */
  failureThreshold: number;
  /** Time in milliseconds to wait before transitioning from OPEN to HALF_OPEN. */
  resetTimeoutMs: number;
  /** Maximum number of trial requests allowed in HALF_OPEN state. */
  halfOpenMaxAttempts: number;
}

// ─── Per-Provider Default Configs ─────────────────────────────────────────────

/**
 * Default circuit breaker configurations per integration provider.
 * These can be overridden when creating a CircuitBreaker instance.
 */
export const PROVIDER_CIRCUIT_CONFIGS: Record<string, CircuitBreakerConfig> = {
  HUBSPOT: {
    failureThreshold: 5,
    resetTimeoutMs: 30_000,
    halfOpenMaxAttempts: 2,
  },
  SALESFORCE: {
    failureThreshold: 5,
    resetTimeoutMs: 60_000,
    halfOpenMaxAttempts: 2,
  },
  GMAIL: {
    failureThreshold: 5,
    resetTimeoutMs: 30_000,
    halfOpenMaxAttempts: 3,
  },
  GOOGLE_CALENDAR: {
    failureThreshold: 5,
    resetTimeoutMs: 30_000,
    halfOpenMaxAttempts: 3,
  },
  OUTLOOK: {
    failureThreshold: 5,
    resetTimeoutMs: 45_000,
    halfOpenMaxAttempts: 2,
  },
  SLACK: {
    failureThreshold: 3,
    resetTimeoutMs: 20_000,
    halfOpenMaxAttempts: 2,
  },
  ODOO: {
    failureThreshold: 3,
    resetTimeoutMs: 15_000,
    halfOpenMaxAttempts: 1,
  },
  QUICKBOOKS: {
    failureThreshold: 5,
    resetTimeoutMs: 60_000,
    halfOpenMaxAttempts: 2,
  },
  XERO: {
    failureThreshold: 5,
    resetTimeoutMs: 60_000,
    halfOpenMaxAttempts: 2,
  },
  JIRA: {
    failureThreshold: 5,
    resetTimeoutMs: 30_000,
    halfOpenMaxAttempts: 2,
  },
  SAP: {
    failureThreshold: 3,
    resetTimeoutMs: 60_000,
    halfOpenMaxAttempts: 2,
  },
  OPENAI: {
    failureThreshold: 5,
    resetTimeoutMs: 45_000,
    halfOpenMaxAttempts: 2,
  },
  ANTHROPIC: {
    failureThreshold: 5,
    resetTimeoutMs: 45_000,
    halfOpenMaxAttempts: 2,
  },
  SENDGRID: {
    failureThreshold: 5,
    resetTimeoutMs: 30_000,
    halfOpenMaxAttempts: 2,
  },
  STRIPE: {
    failureThreshold: 3,
    resetTimeoutMs: 60_000,
    halfOpenMaxAttempts: 2,
  },
  SERVICENOW: {
    failureThreshold: 5,
    resetTimeoutMs: 30_000,
    halfOpenMaxAttempts: 2,
  },
};

/**
 * Fallback config used when no provider-specific config is found.
 */
export const DEFAULT_CIRCUIT_CONFIG: CircuitBreakerConfig = {
  failureThreshold: 5,
  resetTimeoutMs: 30_000,
  halfOpenMaxAttempts: 2,
};

// ─── CircuitBreaker Class ────────────────────────────────────────────────────

/**
 * Circuit Breaker pattern implementation with per-provider configuration
 * and half-open max attempt tracking.
 *
 * State transitions:
 * ```
 *   CLOSED ──(N failures)──> OPEN ──(timeout)──> HALF_OPEN
 *                                                    │
 *                                    success ─> CLOSED
 *                                    failure  ─> OPEN
 *                     (halfOpenMaxAttempts exhausted) ─> OPEN
 * ```
 *
 * @example
 * ```ts
 * // Use provider-specific defaults
 * const cb = CircuitBreaker.forProvider('HUBSPOT');
 * const result = await cb.execute(() => callHubSpotApi());
 *
 * // Use custom config
 * const cb2 = new CircuitBreaker({ failureThreshold: 3, resetTimeoutMs: 10000, halfOpenMaxAttempts: 1 });
 * ```
 */
export class CircuitBreaker {
  private state: CircuitState = CircuitState.CLOSED;
  private failureCount = 0;
  private halfOpenAttempts = 0;
  private lastFailureTime: number | null = null;

  private readonly config: CircuitBreakerConfig;

  constructor(config?: Partial<CircuitBreakerConfig>) {
    this.config = { ...DEFAULT_CIRCUIT_CONFIG, ...config };
  }

  /**
   * Factory: creates a CircuitBreaker using the per-provider default config.
   * Falls back to DEFAULT_CIRCUIT_CONFIG if the provider is not in the map.
   */
  static forProvider(provider: string): CircuitBreaker {
    const providerConfig = PROVIDER_CIRCUIT_CONFIGS[provider];
    return new CircuitBreaker(providerConfig ?? DEFAULT_CIRCUIT_CONFIG);
  }

  /** Returns the current circuit state. */
  getState(): CircuitState {
    return this.state;
  }

  /** Returns the current consecutive failure count. */
  getFailureCount(): number {
    return this.failureCount;
  }

  /** Returns the active configuration. */
  getConfig(): Readonly<CircuitBreakerConfig> {
    return this.config;
  }

  /**
   * Executes the given async function through the circuit breaker.
   *
   * - CLOSED: normal execution; failures are counted.
   * - OPEN: rejects immediately with CircuitBreakerOpenError, unless the
   *   reset timeout has elapsed (transitions to HALF_OPEN).
   * - HALF_OPEN: allows up to `halfOpenMaxAttempts` trial requests. A success
   *   resets to CLOSED; a failure or exceeding max attempts reopens the circuit.
   *
   * @param fn - The async function to execute
   * @returns The result of `fn`
   * @throws CircuitBreakerOpenError when the circuit is OPEN
   * @throws The original error from `fn` when it fails
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    if (this.state === CircuitState.OPEN) {
      if (this.shouldAttemptReset()) {
        this.state = CircuitState.HALF_OPEN;
        this.halfOpenAttempts = 0;
      } else {
        throw new CircuitBreakerOpenError(
          `Circuit breaker is OPEN. Retry after ${this.config.resetTimeoutMs}ms.`,
        );
      }
    }

    // In HALF_OPEN, enforce max attempts
    if (this.state === CircuitState.HALF_OPEN) {
      if (this.halfOpenAttempts >= this.config.halfOpenMaxAttempts) {
        this.state = CircuitState.OPEN;
        this.lastFailureTime = Date.now();
        throw new CircuitBreakerOpenError(
          `Circuit breaker moved to OPEN after exhausting ${this.config.halfOpenMaxAttempts} half-open attempt(s).`,
        );
      }
      this.halfOpenAttempts++;
    }

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure();
      throw error;
    }
  }

  /** Manually reset the circuit breaker to CLOSED state. */
  reset(): void {
    this.state = CircuitState.CLOSED;
    this.failureCount = 0;
    this.halfOpenAttempts = 0;
    this.lastFailureTime = null;
  }

  // ─── Private ──────────────────────────────────────────────────────────────

  private shouldAttemptReset(): boolean {
    if (this.lastFailureTime === null) return false;
    return Date.now() - this.lastFailureTime >= this.config.resetTimeoutMs;
  }

  private onSuccess(): void {
    this.failureCount = 0;
    this.halfOpenAttempts = 0;
    this.state = CircuitState.CLOSED;
  }

  private onFailure(): void {
    this.failureCount++;
    this.lastFailureTime = Date.now();

    if (
      this.failureCount >= this.config.failureThreshold ||
      this.state === CircuitState.HALF_OPEN
    ) {
      this.state = CircuitState.OPEN;
    }
  }
}

// ─── Error ──────────────────────────────────────────────────────────────────

export class CircuitBreakerOpenError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CircuitBreakerOpenError';
  }
}

// ─── Registry ────────────────────────────────────────────────────────────────

/**
 * Singleton registry that lazily creates and caches one CircuitBreaker
 * per provider key. Use this instead of manually creating breakers
 * to ensure a single shared instance per integration.
 *
 * @example
 * ```ts
 * const registry = CircuitBreakerRegistry.getInstance();
 * const result = await registry.getBreaker('HUBSPOT').execute(() => callApi());
 * ```
 */
export class CircuitBreakerRegistry {
  private static instance: CircuitBreakerRegistry | null = null;
  private readonly breakers = new Map<string, CircuitBreaker>();

  private constructor() {}

  static getInstance(): CircuitBreakerRegistry {
    if (!CircuitBreakerRegistry.instance) {
      CircuitBreakerRegistry.instance = new CircuitBreakerRegistry();
    }
    return CircuitBreakerRegistry.instance;
  }

  /**
   * Get or create a CircuitBreaker for the given provider.
   * Uses the provider-specific config from PROVIDER_CIRCUIT_CONFIGS.
   */
  getBreaker(provider: string): CircuitBreaker {
    const key = provider.toUpperCase();
    let breaker = this.breakers.get(key);
    if (!breaker) {
      breaker = CircuitBreaker.forProvider(key);
      this.breakers.set(key, breaker);
    }
    return breaker;
  }

  /**
   * Returns the state of all registered circuit breakers.
   * Useful for health check endpoints and monitoring.
   */
  getStatus(): Record<string, { state: CircuitState; failureCount: number }> {
    const status: Record<string, { state: CircuitState; failureCount: number }> = {};
    for (const [key, breaker] of this.breakers) {
      status[key] = {
        state: breaker.getState(),
        failureCount: breaker.getFailureCount(),
      };
    }
    return status;
  }

  /** Reset all circuit breakers to CLOSED state. */
  resetAll(): void {
    for (const breaker of this.breakers.values()) {
      breaker.reset();
    }
  }
}

File: ai-runtime/orchestrator/circuit_breaker.py
"""Circuit Breaker pattern for AI Runtime.

Provides per-provider circuit breakers that prevent cascading failures
by tracking consecutive errors and temporarily blocking calls to
failing providers.
"""

import asyncio
import logging
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, TypeVar

logger = logging.getLogger(__name__)

T = TypeVar("T")


class CircuitState(str, Enum):
    CLOSED = "CLOSED"
    OPEN = "OPEN"
    HALF_OPEN = "HALF_OPEN"


@dataclass
class CircuitBreakerConfig:
    failure_threshold: int = 5
    reset_timeout_s: float = 30.0
    half_open_max_attempts: int = 2


# Per-provider configs matching P6 spec Section 3
PROVIDER_CONFIGS: dict[str, CircuitBreakerConfig] = {
    "OPENAI": CircuitBreakerConfig(failure_threshold=5, reset_timeout_s=45.0, half_open_max_attempts=2),
    "ANTHROPIC": CircuitBreakerConfig(failure_threshold=5, reset_timeout_s=45.0, half_open_max_attempts=2),
    "AZURE_OPENAI": CircuitBreakerConfig(failure_threshold=5, reset_timeout_s=45.0, half_open_max_attempts=2),
}

DEFAULT_CONFIG = CircuitBreakerConfig()


class CircuitBreakerOpenError(Exception):
    """Raised when circuit breaker is OPEN and rejecting calls."""
    pass


class CircuitBreaker:
    """Circuit breaker with CLOSED -> OPEN -> HALF_OPEN state machine."""

    def __init__(self, config: CircuitBreakerConfig | None = None):
        self._config = config or DEFAULT_CONFIG
        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._half_open_attempts = 0
        self._last_failure_time: float | None = None

    @classmethod
    def for_provider(cls, provider: str) -> "CircuitBreaker":
        config = PROVIDER_CONFIGS.get(provider, DEFAULT_CONFIG)
        return cls(config)

    @property
    def state(self) -> CircuitState:
        return self._state

    @property
    def failure_count(self) -> int:
        return self._failure_count

    async def execute(self, fn: Callable[..., Any], *args: Any, **kwargs: Any) -> Any:
        """Execute fn through the circuit breaker."""
        if self._state == CircuitState.OPEN:
            if self._should_attempt_reset():
                self._state = CircuitState.HALF_OPEN
                self._half_open_attempts = 0
                logger.info("Circuit breaker transitioning to HALF_OPEN")
            else:
                raise CircuitBreakerOpenError(
                    f"Circuit breaker is OPEN. Retry after {self._config.reset_timeout_s}s."
                )

        if self._state == CircuitState.HALF_OPEN:
            if self._half_open_attempts >= self._config.half_open_max_attempts:
                self._state = CircuitState.OPEN
                self._last_failure_time = time.monotonic()
                raise CircuitBreakerOpenError(
                    f"Circuit breaker moved to OPEN after exhausting "
                    f"{self._config.half_open_max_attempts} half-open attempt(s)."
                )
            self._half_open_attempts += 1

        try:
            result = await fn(*args, **kwargs)
            self._on_success()
            return result
        except Exception as exc:
            self._on_failure()
            raise

    def reset(self) -> None:
        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._half_open_attempts = 0
        self._last_failure_time = None

    def _should_attempt_reset(self) -> bool:
        if self._last_failure_time is None:
            return False
        return (time.monotonic() - self._last_failure_time) >= self._config.reset_timeout_s

    def _on_success(self) -> None:
        self._failure_count = 0
        self._half_open_attempts = 0
        self._state = CircuitState.CLOSED

    def _on_failure(self) -> None:
        self._failure_count += 1
        self._last_failure_time = time.monotonic()
        if (
            self._failure_count >= self._config.failure_threshold
            or self._state == CircuitState.HALF_OPEN
        ):
            self._state = CircuitState.OPEN
            logger.warning(
                "Circuit breaker OPEN after %d failures", self._failure_count
            )
